primitives/src/merkle_tree/light_weight.rs

```diff
@@ -0,0 +1,195 @@
+};
+use crate::{errors::PrimitivesError, impl_forgetable_merkle_tree_scheme, impl_merkle_tree_scheme};
+use ark_std::{
+    borrow::Borrow, boxed::Box, fmt::Debug, marker::PhantomData, string::ToString, vec, vec::Vec,
+};
+use num_bigint::BigUint;
+use num_traits::pow::pow;
+use serde::{Deserialize, Serialize};
+use typenum::Unsigned;
+
+impl_merkle_tree_scheme!(LightWeightMerkleTree, build_light_weight_tree_internal);
+impl_forgetable_merkle_tree_scheme!(LightWeightMerkleTree);
+
+impl<E, H, I, Arity, T> AppendableMerkleTreeScheme for LightWeightMerkleTree<E, H, I, Arity, T>
+where
+    E: Element,
+    H: DigestAlgorithm<E, I, T>,
+    I: Index + From<u64> + AddAssign + ToTraversalPath<Arity>,
+    Arity: Unsigned,
+    T: NodeValue,
+{
+    fn push(&mut self, elem: impl Borrow<Self::Element>) -> Result<(), PrimitivesError> {
+        <Self as AppendableMerkleTreeScheme>::extend(self, [elem])
+    }
+
+    fn extend(
+        &mut self,
+        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
+    ) -> Result<(), PrimitivesError> {
+        let mut iter = elems.into_iter().peekable();
+
+        let traversal_path = I::from(self.num_leaves).to_traversal_path(self.height);
+        self.num_leaves += self.root.extend_and_forget_internal::<H, Arity>(
+            self.height,
+            &I::from(self.num_leaves),
+            &traversal_path,
+            true,
+            &mut iter,
+        )?;
+        if iter.peek().is_some() {
+            return Err(PrimitivesError::ParameterError(
+                "Exceed merkle tree capacity".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod mt_tests {
+    use crate::{
+        merkle_tree::{
+            internal::{MerkleNode, MerkleProof},
+            prelude::{RescueLightWeightMerkleTree, RescueMerkleTree},
+            *,
+        },
+        rescue::RescueParameter,
+    };
+    use ark_ed_on_bls12_377::Fq as Fq377;
+    use ark_ed_on_bls12_381::Fq as Fq381;
+    use ark_ed_on_bn254::Fq as Fq254;
+
+    #[test]
+    fn test_light_mt_builder() {
+        test_light_mt_builder_helper::<Fq254>();
+        test_light_mt_builder_helper::<Fq377>();
+        test_light_mt_builder_helper::<Fq381>();
+    }
+
+    fn test_light_mt_builder_helper<F: RescueParameter>() {
+        assert!(RescueLightWeightMerkleTree::<F>::from_elems(1, &[F::from(0u64); 3]).is_ok());
+        assert!(RescueLightWeightMerkleTree::<F>::from_elems(1, &[F::from(0u64); 4]).is_err());
+    }
+
+    #[test]
+    fn test_light_mt_insertion() {
+        test_light_mt_insertion_helper::<Fq254>();
+        test_light_mt_insertion_helper::<Fq377>();
+        test_light_mt_insertion_helper::<Fq381>();
+    }
+
+    fn test_light_mt_insertion_helper<F: RescueParameter>() {
+        let mut mt = RescueLightWeightMerkleTree::<F>::from_elems(2, &[]).unwrap();
+        assert_eq!(mt.capacity(), BigUint::from(9u64));
```

Comment:
Is `capacity()` a BigUint?

Reply:
Yes. It's from the trait definition of `MerkleTree`.